### PR TITLE
Ensure the MaterialApp's navigatorObservers list not not null before creating WidgetApp.

### LINKED ISSUE #18904: Creating a MaterialApp with a custom Navigator fails because of assertions
By specifying the `builder` of a `MaterialApp` widget, it should be possible to use a custom `Navigator` (or no navigator at all. The assertions of the ``WidgetsApp`` constructor make it impossible.

Error:

```
I/flutter (19749): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
I/flutter (19749): The following assertion was thrown building MaterialApp(dirty, state: _MaterialAppState#f5965):
I/flutter (19749): 'package:flutter/src/widgets/app.dart': Failed assertion: line 93 pos 15: 'navigatorObservers !=
I/flutter (19749): null': is not true.
```

Removing these 2 assertions fixes the problem:
```dart
assert(navigatorObservers != null),
assert(onGenerateRoute != null || navigatorObservers == const <NavigatorObserver>[]),
```

Example App:

```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

final Map<String, WidgetBuilder> routes = {
  '/': (context) => HomePage(),
  '/credits': (context) => CreditsPage(),
};

RouteFactory customOnGenerateRoute = (RouteSettings settings) {
  var name = settings.name;
  print('going to route $name');
  return new MaterialPageRoute(
    builder: routes[name],
    settings: settings,
  );
};

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Navigator Demo',
      theme: new ThemeData(primarySwatch: Colors.blue),
      builder: (BuildContext context, Widget child) {
        return Navigator(
          onGenerateRoute: customOnGenerateRoute,
        );
      },
    );
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Home')),
      body: Center(
        child: RaisedButton(
          child: Text('Credits'),
          onPressed: () {
            Navigator.of(context).pushNamed('/credits');
          },
        ),
      ),
    );
  }
}

class CreditsPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Credits')),
      body: Center(
        child: Text('made by boformer'),
      ),
    );
  }
}
```

### PR DESCRIPTION
**Pull request by ghost**
## Description

When the MaterialApp was creating the underlying WidgetApp when there was a custom navigator and no routes set, it was passing in `null` for the `navigatorObservers` parameter. This would trip an assertion in the WidgetApp as it expects an empty observers list if there are no routes. 

This PR fixes that and adds a regression test.

## Related Issues

Fixes #18904 

## Tests

I added the following tests:

material/app_test - 'custom navigator with without routes supported'

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/material/app_test.dart [hunk 1]
```diff
@@ -381,6 +381,23 @@ void main() {
...
   });
 
+  testWidgets('custom navigator with without routes supported', (WidgetTester tester) async {
+    // Regression test for https://github.com/flutter/flutter/issues/18904
+    await tester.pumpWidget(
+      MaterialApp(
+        builder: (BuildContext context, Widget child) {
+          return Navigator(
+            onGenerateRoute: (RouteSettings settings) {
+              return MaterialPageRoute<void>(builder: (BuildContext context) => Scaffold());
```
Comment (HansMuller) [inline packages/flutter/test/material/app_test.dart]: const Scaffold()

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/test/material/app_test.dart (hunk 1 above)
Comment (DarrenGoog) [reply packages/flutter/test/material/app_test.dart]: Done. thx.

File: packages/flutter/test/material/app_test.dart [hunk 2]
```diff
@@ -381,6 +381,23 @@ void main() {
     expect(log, <String>['onGenerateRoute /', 'onUnknownRoute /']);
   });
 
+  testWidgets('custom navigator with without routes supported', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/material/app_test.dart]: Nit: with or without?

Comment (DarrenGoog) [reply packages/flutter/test/material/app_test.dart]: D'oh. Changed my mind mid stream and my hands didn't get the memo. Thanks for catching it.

**Comment by DarrenGoog** [comment]
I totally missed the PR from @rmtmckenzie: **https://github.com/flutter/flutter/pull/29413* that solved this in the same way. Closing this PR and will use his.

**Pull request closed without merging**